pkg/gardenlet/controller/extensions/extensions_sync_control.go

```diff
@@ -0,0 +1,138 @@
+
+package extensions
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	gardenextensionsclient "github.com/gardener/gardener/pkg/client/extensions/clientset/versioned"
+	gardenextensionsinformers "github.com/gardener/gardener/pkg/client/extensions/informers/externalversions"
+	"github.com/gardener/gardener/pkg/client/extensions/listers/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// ExtensionStateSyncController replicates the extensions' resources states in Shoot's ShootState resource
+type ExtensionStateSyncController struct {
+	k8sGardenClient      kubernetes.Interface
+	seedClient           kubernetes.Interface
+	extensionsInformers  gardenextensionsinformers.SharedInformerFactory
+	extensionsSeedClient gardenextensionsclient.Interface
+
+	context             context.Context
+	cancelFunc          context.CancelFunc
+	log                 *logrus.Entry
+	infrastructuresList v1alpha1.InfrastructureLister
+	workersList         v1alpha1.WorkerLister
+
+	recorder record.EventRecorder
+
+	infrastructuresHasSynced cache.InformerSynced
+	workerHasSynced          cache.InformerSynced
+
+	infrastructureExtensionQueue workqueue.RateLimitingInterface
+	workerExtensionQueue         workqueue.RateLimitingInterface
+
+	waitGroup              sync.WaitGroup
+	numberOfRunningWorkers int
+	workerCh               chan int
+	informerShutdown       chan struct{}
+}
+
+// NewExtensionsSyncController creates new controller that syncs extensions states to ShootSnapshot
+func NewExtensionsSyncController(parentCtx context.Context, gardenClient, seedClient kubernetes.Interface, extensionsSeedClient gardenextensionsclient.Interface, extensionsInformers gardenextensionsinformers.SharedInformerFactory, config *config.GardenletConfiguration, log *logrus.Entry, recorder record.EventRecorder) *ExtensionStateSyncController {
+	var (
+		infraInformer        = extensionsInformers.Extensions().V1alpha1().Infrastructures()
+		workerInformer       = extensionsInformers.Extensions().V1alpha1().Workers()
+		infrastructuresList  = infraInformer.Lister()
+		workersList          = workerInformer.Lister()
+		childCtx, cancelFunc = context.WithCancel(parentCtx)
+	)
+
+	extensionController := &ExtensionStateSyncController{
+		k8sGardenClient:              gardenClient,
+		seedClient:                   seedClient,
+		extensionsSeedClient:         extensionsSeedClient,
+		extensionsInformers:          extensionsInformers,
+		infrastructuresList:          infrastructuresList,
+		log:                          log,
+		workerExtensionQueue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "worker-extension-controllers"),
```

Comment:
Let's create a Workers struct which will contain all the necessary worker queues and will also create all the necessary workers later on in the `Run` method.